transactions/read-only/parallel.md

```diff
@@ -0,0 +1,215 @@
+    B -->|no| A
+    A --> D[write_window_deadline passed?]
+    D -->|yes| F[read-only trx queue empty?]
+    D -->|no| A
+    F -->|yes| A
+    F -->|no| R
+    
+    R --> S[read-only trx queue empty?]
+    S -->|yes| A
+    S -->|no| R
+    R --> T[read_window_deadline passed?]
+    T -->|yes| A
+    T -->|no| R  
+```
+
+### Handling main thread functions that are not safe to read-only transaction execution in `read` Window
+Several options are considered to handle 
+- Drop the functions. This in not acceptable as it changes the behavior.
+- Re-post to the main thread. In `read` window, re-post a not-read-only-safe function back to the main thread. This is simple to implement but breaks the order of functions and waste time in popping and pushing. 
+- Modify appbase `execution_priority_queue` by adding a function type with value `read-only-safe` or `not-read-only-safe`. In `write` window, everything works as it is now. In `read` window, only `thread-safe` requests are dequeued and processed. This avoids introducing new queues and keeps the order of the request. But as the queue is a priority queue based on function priority, it is infeasible to incorporate the additional type into the priority in a single queue.
+- Separate appbase queue into two queues: `read-only-safe` or `not-read-only-safe`. The function type is added when a function is posted. 
+  - In `read` window, only functions in `read-only-safe` queue are executed.
+  - In `write` window, compare the priorities of the top functions in `read-only-safe` and `not-read-only-safe`. The one with higher priority is executed. If tied, three options are considered:
+    - `not-read-only-safe` function is favored 
+    - randomly pick one
+    - add a time attribute to the functions and the older one is picked. This keeps the original behavior. Even though at a cost of the extra time field and an extra comparison, this option seems best.
+
+### Read-only Transaction Priority Queue
+The queue is used to store read-only transactions during `write` window. To maintain the time order when a transaction is put back into the queue for the next round due to `red` window deadline, a priority queue based on the first time when a transaction is queued is used.
+```c++
+struct read_only_trx {
+   fc::time_point          initial_queued_time;
+   packed_transaction_ptr  trx;
+   next_func_t             next;
+};
+      
+struct read_only_trx_less {
+   bool operator() (const read_only_trx& a, const read_only_trx& b) {
+      // earlier queued time has higher priority
+      return a.initial_queued_time > b.initial_queued_time;
+   }
+};
+
+std::priority_queue<read_only_trx, std::deque<read_only_trx>, read_only_trx_less>  read_only_trx_queue;     
+```
+
+### Configuration Options
+
+- `read-only-num-threads`: the number of threads in read-only transaction execution thread pool. Default to `0`. If it is `0`, read-only transactions are executed on the main thread sequentially as they arrive
```

Comment:
This is really the number of threads in the read-only *operations* thread pool. Seems like we should distinguish in this document between read-only transactions and read-only operations. In a lot of cases we mean read-only operations not read-only transactions.

read-only transaction - an Antelope transaction that does not modify state
read-only operation - a task (like get_info) that does not modify state

Reply:
Sorry for the confusions. I tried to use read functions but not consistently. I will define read-only transaction and read-only operation at the beginning of the document and use them consistently.